source/slang/slang.cpp

```diff
@@ -3708,6 +3825,48 @@ CompositeComponentType::CompositeComponentType(
     }
 }
 
+SlangResult CompositeComponentType::computeDependencyBasedHash(slang::Checksum* outHashCode)
+{
+    auto componentCount = getChildComponentCount();
+
+    slang::Checksum hashCode;
+    MD5HashGen hashGen;
+    MD5Context context;
+    hashGen.init(&context);
+
+    for (Index i = 0; i < componentCount; ++i)
+    {
+        slang::Checksum tempHash;
+        getChildComponent(i)->computeDependencyBasedHash(&tempHash);
+        hashGen.update(&context, tempHash);
```

Comment:
There is a "code smell" here that is making me realize the signatures of these `compute...Hash(...)` functions aren't quite what we want. The functions should probably *not* be yielding a `Checksum` as their output, and should instead be functions that take an `MD5HashGen` and `MD5Context` as input (or a convenience type that bundles the two...):

```
SlangResult CompositeComponentType::computeDependencyBasedHash(MD5HashGen& hashGen, MD5Context const* context)
{ ... }
```

Using that as the signature for the core recursive routine would yield two big improvements:

1. Most of the implementation routines would no longer need to declare their own local `hashCode`, `hashGen`, and `context` variables, nor call `hashGen.init()`.

2. When making a recursive call like the one tagged for this comment, a function would just be able to pass `hashGen` and `context` down to the callee, rather than have to declare a distinct `tempHash` and then use `hashGen.update(..., tempHash)`.

Each of these leads to less code overall (so that's a nice win already), but item (2) there could actually lead to a better quality hashed output for hashing algorithms where the internal state of a hash generator is larger than the size of the hash code / digest produced at the end.

Reply:
I'm realizing just now that the call to compute the hash for the child component isn't doing anything on a failure `slang::Result`. That seems incorrect to me.

Either these routines should be using some kind of `SLANG_RETURN_ON_FAIL` or similar to handle the recursive calls, so that errors get propagated, or if none of these routines can actually fail, we shouldn't be defining them to return a `Result` in the first place.

Reply:
Let's create a class `MD5DigestBuilder` that owns both a `MD5HashGen` and `MD5Context`, then 
define the `computeDependencyBasedHashImpl(MD5DigestBuilder& builder)` as an abstract method in `ComponentTypeBase`, and implement the existing `IComponentType::computeDependencyBasedHash()` with a direct call to `computeDependencyBasedHashImpl` with a fresh  `MD5DigestBuilder`, there we have the opportunity to call `Linkage::computeDependencyBasedHash()` once to initialize the `MD5DigestBuilder`.

The rest of implementations of `computeDependencyBasedHashImpl` should only recursively call `computeDependencyBasedHashImpl` and not the public `computeDependencyBasedHash`.

Reply:
Restructuring done and should be working based on shader cache tests, will be in next push